compiler/noirc_frontend/src/node_interner.rs

```diff
@@ -1119,9 +1174,29 @@ impl NodeInterner {
     /// Returns what the next trait impl id is expected to be.
     /// Note that this does not actually reserve the slot so care should
     /// be taken that the next trait impl added matches this ID.
-    pub(crate) fn next_trait_impl_id(&self) -> TraitImplId {
+    pub fn next_trait_impl_id(&self) -> TraitImplId {
         TraitImplId(self.trait_implementations.len())
     }
+
+    /// Removes all TraitImplKind::Assumed from the list of known impls for the given trait
+    pub fn remove_assumed_trait_implementations_for_trait(&mut self, trait_id: TraitId) {
+        let entries = self.trait_implementation_map.entry(trait_id).or_default();
+        entries.retain(|(_, kind)| matches!(kind, TraitImplKind::Normal(_)));
+    }
+
+    /// Tags the given identifier with the selected trait_impl so that monomorphization
+    /// can later recover which impl was selected, or alternatively see if it needs to
+    /// decide which (because the impl was Assumed).
```

Comment:
decide which impl to select?

Reply:
Right. In cases where we only have an assumed impl (e.g. because of an `where T: Eq`, we assume there is some Eq impl for T, and the type system later verifies this when the function is called but we can't actually "select" the impl there because we can't change the contents of the function that was called. So we rely on monomorphization to later select the correct impl. This is also why we need to follow the coherence restriction such that there is only ever at most 1 applicable impl. Since there's only 1 impl that will match, we can rely on monomorphization to re-select the same impl.